reana_job_controller/job_manager.py

```diff
@@ -0,0 +1,131 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2019 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Job Manager."""
+
+import json
+import shlex
+
+from reana_commons.utils import calculate_file_access_time
+
+from reana_db.database import Session
+from reana_db.models import JobStatus, Workflow, JobCache
+from reana_db.models import Job as JobTable
+
+from reana_job_controller.config import MAX_JOB_RESTARTS
+
+
+class JobManager():
+    """Job management interface."""
+
+    def __init__(self, docker_img='', cmd=[], env_vars={}, job_id=None,
+                 workflow_uuid=None):
+        """Instanciates basic job.
+
+        :param docker_img: Docker image.
+        :type docker_img: str
+        :param cmd: Command to execute.
+        :type cmd: list
+        :param env_vars: Environment variables.
+        :type env_vars: dict
+        :param job_id: Unique job id.
+        :type job_id: str
+        :param workflow_id: Unique workflow id.
+        :type workflow_id: str
+        """
+        self.docker_img = docker_img or ''
+        if isinstance(cmd, str):
+            self.cmd = shlex.split(cmd)
+        else:
+            self.cmd = cmd or []
+        self.env_vars = env_vars or {}
+        self.job_id = job_id
+        self.workflow_uuid = workflow_uuid
+
+    def execution_hook(fn):
+        """Add before execution hooks and DB operations."""
+        def wrapper(inst, *args, **kwargs):
+            inst.before_execution()
+            backend_job_id = fn(inst, *args, **kwargs)
+            inst.create_job_in_db(backend_job_id)
+            inst.cache_job()
+            return backend_job_id
+        return wrapper
+
+    def before_execution(self):
+        """Before job submission hook."""
```

Comment:
Note that @lukasheinrich was trying out git-base syncing of needed input/output files instead of rsync-based syncing